Redesign Task Requestor Details Modal on Task Request Page and add Feature Flag

Hariom01010 opened the pull request:
<!-- Date Here in dd mmm yyyy-->
Date: 24 Feb 2024
<!--Developer Name Here-->
Developer Name: @Hariom01010 

---

## Issue Ticket Number
<!--Issue ticket this PR closes-->
- Closes #803 
## Description
This PR aims at redesigning the task request details modal and adding it under feature flag.
Task Requestore Modal [Design File](https://www.figma.com/design/sricuOxvTsVTlB1IoZYAyP/Task-Requestor-Modal?node-id=33-4&t=CFTiIO3XnfYfSDmz-1)
Task Requestore Modal [Design Doc](https://docs.google.com/document/d/1PIsvHdunQry-ty3TUKzwotdcYPqKxoo2eHYibZsmSnA/edit?usp=sharing)

<!--Description of the changes made in this PR-->

### Documentation Updated?

- [ ] Yes
- [X] No

<!--Additional notes about documentation update if applicable-->

### Under Feature Flag

- [X] Yes
- [ ] No

<!--Indicate if changes are under a feature flag-->

### Database Changes

- [ ] Yes
- [X] No

<!--Notes on any database changes-->

### Breaking Changes

- [ ] Yes
- [X] No

<!--Notes on breaking changes or related issue tickets if applicable-->

### Development Tested?

- [X] Yes
- [ ] No

<!--Confirmation of local testing during development-->

## Screenshots
<details>
<summary>Screenshot 1</summary>

<!-- Attach your screenshots here👇-->
![Screenshot 2025-02-28 173357](https://github.com/user-attachments/assets/535c65bf-ed74-4563-8cc0-3fb8a11fe2bf)
![Screenshot 2025-02-28 173622](https://github.com/user-attachments/assets/60e74ebe-8138-4541-bab1-9ca8b3e70a5e)
![image](https://github.com/user-attachments/assets/660f511a-7bf4-4d22-9182-ebd367b230be)

https://github.com/user-attachments/assets/f8858a66-1138-460e-a14e-9fcd53d42f3d


</details>

<!--Attach or link to relevant screenshots or visual aids-->




## Test Coverage
<details>
<summary>Screenshot 1</summary>

<!-- Attach your screenshots here👇-->
Task Request Test Coverage:
![task request](https://github.com/user-attachments/assets/c5694b2a-4494-4413-8486-067cff6df7fe)

Task Request Details Test Coverage:
![TASK REQUEST DETAILS TEST](https://github.com/user-attachments/assets/bbfa828d-46fd-4881-ab08-5036e4f8d651)

</details>

<!--Attach Details on test coverage and outcomes-->

## Additional Notes

<!--Any additional notes, considerations or explanations for reviewers-->
The task is under feature flag and after it has been tested on prod for 10-15 days and everything is working fine. Then a new task/PR will be created to remove the feature flag

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **New Features**
  - Enhanced the task request details modal to display information more consistently, including improved fallback text for missing details based on the development environment.
  
- **Style**
  - Introduced a new CSS variable and classes for the modal's appearance, optimizing responsiveness and visual consistency across devices.

- **Tests**
  - Added new test cases to verify the rendering of different modals based on the development environment flag, including handling of long descriptions.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

RishiChaubey31 commented:
Can you please also add a screenshot of the mobile view in the screenshot section

RishiChaubey31 commented on task-requests/details/script.js:
```diff
@@ -620,6 +620,11 @@ function populateModalContent(index) {
   }
   const modal = document.getElementById('requestor_details_modal_content');
+  if (params.get('dev') == 'true') {
+    console.log(modal);
```
Is this log necessary

MayankBansal12 replied:
nit: remove `console.log`

MayankBansal12 commented on task-requests/details/script.js:
```diff
@@ -683,23 +709,62 @@ function populateModalContent(index) {
     descriptionValue.innerHTML = html;
     descriptionValue.className = 'requestor_description_details';
   } else {
-    descriptionValue.textContent = userData.description;
+    if (params.get('dev') == 'true') {
+      if (userData.description) {
+        descriptionValue.textContent = userData.description;
+      } else {
+        descriptionValue.textContent = 'N/A';
+        descriptionValue.className = 'proposed-description-value';
+      }
+    } else {
+      descriptionValue.textContent = userData.description;
+    }
   }
```
please refactor this

VaibhavSingh8 commented on task-requests/details/script.js:
> why use `params.get('dev') == 'true'` everywhere? check if a variable is already storing this otherwise make one and use that

@Hariom01010 Do something like this:

```js
const isDev = urlParams.get('dev') === 'true';
```

And use this isDev where you need to check.

VaibhavSingh8 commented:
@Hariom01010 Couldn't find the design file and design doc link in the PR description. Can you please update it?

Also, was the design provided by the designer or you made it?

Hariom01010 commented:
> @Hariom01010 Couldn't find the design file and design doc link in the PR description. Can you please update it?
> 
> Also, was the design provided by the designer or you made it?

@VaibhavSingh8 I have updated the PR Description and added the design file and design doc. The initial design was provided by @AnujChhikara which was reviewed by design team and changes were made as needed

VaibhavSingh8 commented:
> > @Hariom01010 Couldn't find the design file and design doc link in the PR description. Can you please update it?
> > Also, was the design provided by the designer or you made it?
> 
> @VaibhavSingh8 I have updated the PR Description and added the design file and design doc. The initial design was provided by @AnujChhikara which was reviewed by the design team and changes were made as needed

Cool! I can see Prakash sir's approval in the comments. Please change the design doc status to approved.

Hariom01010 commented:
> > > @Hariom01010 Couldn't find the design file and design doc link in the PR description. Can you please update it?
> > > Also, was the design provided by the designer or you made it?
> > 
> > @VaibhavSingh8 I have updated the PR Description and added the design file and design doc. The initial design was provided by @AnujChhikara which was reviewed by the design team and changes were made as needed
> 
> Cool! I can see Prakash sir's approval in the comments. Please change the design doc status to approved.

Updated the status. Thanks for noticing it!

AnujChhikara commented on task-requests/details/script.js:
```diff
@@ -683,23 +708,58 @@ function populateModalContent(index) {
-  modalContent.appendChild(descriptionText);
-  modalContent.appendChild(descriptionValue);
+
+  if (isDev) {
```
Instead of using the dev flag in multiple places, can we create a separate populateModalContent function? Based on isDev, we can render the appropriate version, making it easier to remove the dev flag in the future.

AnujChhikara commented on task-requests/details/style.css:
```diff
@@ -555,6 +534,30 @@ table tr td {
+.proposed-description-text {
+  font-weight: 500;
+}
+.proposed-start-date-value,
```
Ensure all class names follow a consistent naming convention. Use either underscores (_) or hyphens (-), aligning with the existing class names in the codebase.

AnujChhikara reviewed (changes requested):
Please write a test to verify the behavior of the dev flag.

RishiChaubey31 commented on task-requests/details/style.css:
```diff
@@ -570,3 +573,61 @@ p:hover {
+    line-height: 1.75rem;
+  }
+  .taskRequest__title__subtitle {
+    font-size: 0.875rem;
```
nit: make sure all values are in rem other than borders or media queries and some specific values and if the rem values are up to 3 decimal places  you can round them off

other than that 

LGTM 👍

VaibhavSingh8 commented:
LGTM! My comments have been addressed. Waiting for other reviewers to resolve their concerns.

MayankBansal12 commented on __tests__/task-requests/task-requestDetails.test.js:
```diff
@@ -160,6 +160,42 @@ describe('Task request details page', () => {
     expect(rejectButton).toBeTruthy();
   });
+
+  it('Should render new modal if isDev flag is enabled and old modal is rendered if isDev flag is disabled', async function () {
```
please break down this test into two parts -
- when feature flag is enabled
- when feature flag is disabled

this makes it easier to remove the test when removing feature flag

AnujChhikara replied:
Can we please divide this test into two separate tests

MayankBansal12 commented on task-requests/details/style.css:
```diff
@@ -555,6 +534,30 @@ table tr td {
+  margin: 1rem 0rem;
+}
+
+.proposed_start-date_text,
+.proposed_end-date_text,
```
```suggestion
.proposed_start_date_text,
.proposed_end_date_text,
```

AnujChhikara commented on task-requests/details/script.js:
```diff
@@ -694,12 +697,68 @@ function populateModalContent(index) {
+    // Proposed Deadline Value: Add new class
+    proposedDeadlineValue.className = 'proposed_end_date_value';
+
+    // Description: Add new text and classname
```
How these comments are useful?

VaibhavSingh8 reviewed (approved):
LGTM

AnujChhikara reviewed (changes requested):
Your modal design does not align with the approved design. Please update it to match the specified requirements.

VaibhavSingh8 commented:
Is the UI screenshot updated after all the changes? @Hariom01010

Hariom01010 commented:
> @Hariom01010 didn't we discuss for adding the screen recording too ??
oh yes thanks, let me attach the updated recording

Achintya-Chatterjee commented on task-requests/details/script.js:
```diff
@@ -694,12 +697,70 @@ function populateModalContent(index) {
+    const modalCloseButton = document.getElementById(
+      'requestor_details_modal_close',
+    );
+    console.log(modalCloseButton);
```
why do we need `console.log` here?

Achintya-Chatterjee commented on __tests__/task-requests/task-requestDetails.test.js:
```diff
@@ -160,6 +160,44 @@ describe('Task request details page', () => {
     expect(rejectButton).toBeTruthy();
   });
+
+  it('Should render new modal if isDev flag is enabled', async function () {
+    await page.goto(
+      `${LOCAL_TEST_PAGE_URL}/task-requests/details/?id=dM5wwD9QsiTzi7eG7Oq5&dev=true`,
+    );
+    await page.waitForNetworkIdle();
```
What happens if `taskRequest.users[index]` is `undefined`? I feel we should write a test for the same

Achintya-Chatterjee commented on __tests__/task-requests/task-requestDetails.test.js:
```diff
@@ -160,6 +160,44 @@ describe('Task request details page', () => {
+    ).not.toBeNull();
+  });
+
+  it('Should render old modal if isDev flag is disabled', async function () {
+    await page.goto(
+      `${LOCAL_TEST_PAGE_URL}/task-requests/details/?id=dM5wwD9QsiTzi7eG7Oq5`,
+    );
+    await page.waitForNetworkIdle();
+    await page.click('.info__more');
+    await page.waitForSelector('#requestor_details_modal_content', {
+      visible: true,
+    });
+
+    expect(await page.$('.requestor_details_modal_content')).not.toBeNull();
```
Does the modal properly handle long descriptions? I don't see any test for it

Hariom01010 commented:
> could you please provide a better PR name @Hariom01010 ?

Updated the PR name
